### PR TITLE
feat: enable sort spilling when memory reach the limit.

### PR DESCRIPTION
**Pull request by RinChanNOWWW**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

- Use builder pattern to build the merge sort pipeline.
- In `TransformSortMerge`, if memory limit is reached, add meta to the output `DataBlock` to let `TransformSortSpill` spill the blocks.

Tracking in #13889

## Tests
- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test  - _Explain why_

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/13989)

### CODE REVIEW

File: tests/sqllogictests/suites/base/03_common/03_0041_sort_spill.test [hunk 1]
```diff
@@ -0,0 +1,402 @@
+statement ok
```
Comment (BohuTANG) [inline tests/sqllogictests/suites/base/03_common/03_0041_sort_spill.test:1]: It seems that we don't check the spill or not in this test?

Comment (BohuTANG) [reply tests/sqllogictests/suites/base/03_common/03_0041_sort_spill.test:1]: We can check the metrics counter.

Comment (RinChanNOWWW) [reply tests/sqllogictests/suites/base/03_common/03_0041_sort_spill.test:1]: Can we get the metrics by SQL command?

Comment (BohuTANG) [reply tests/sqllogictests/suites/base/03_common/03_0041_sort_spill.test:1]: Yes, 
`select metric, value from system.metrics where metric like '%<metric-name>%';`

**Review by zhang2014 — APPROVED** [review]

**Review by sundy-li — APPROVED** [review]

**Pull request merged**